vana/utils/transaction.py

```diff
@@ -0,0 +1,167 @@
+        self.web3 = web3
+        self.account = account
+        self._nonce_cache: Dict[str, int] = {}
+        self._last_nonce_refresh = 0
+        self.nonce_refresh_interval = 60
+        self.chain_id = self.web3.eth.chain_id
+
+    def _get_safe_nonce(self) -> int:
+        """
+        Get the next safe nonce, accounting for pending transactions
+        """
+        current_time = time.time()
+        cache_key = self.account.address
+
+        # Refresh nonce cache if expired
+        if current_time - self._last_nonce_refresh > self.nonce_refresh_interval:
+            # Get the latest confirmed nonce
+            confirmed_nonce = self.web3.eth.get_transaction_count(self.account.address, 'latest')
+            # Get pending nonce
+            pending_nonce = self.web3.eth.get_transaction_count(self.account.address, 'pending')
+            # Use the higher of the two to avoid nonce conflicts
+            self._nonce_cache[cache_key] = max(confirmed_nonce, pending_nonce)
+            self._last_nonce_refresh = current_time
+
+        # Get and increment the cached nonce
+        nonce = self._nonce_cache.get(cache_key, 0)
+        self._nonce_cache[cache_key] = nonce + 1
+        return nonce
+
+    def _clear_pending_transactions(self):
+        """
+        Clear pending transactions by sending zero-value transactions with higher gas price
+        """
+        try:
+            # Get all pending transactions for the account
+            pending_nonce = self.web3.eth.get_transaction_count(self.account.address, 'pending')
+            confirmed_nonce = self.web3.eth.get_transaction_count(self.account.address, 'latest')
+
+            if pending_nonce > confirmed_nonce:
+                vana.logging.info(f"Clearing {pending_nonce - confirmed_nonce} pending transactions")
+
+                # Send replacement transactions with higher gas price
+                for nonce in range(confirmed_nonce, pending_nonce):
+                    replacement_tx = {
+                        'from': self.account.address,
+                        'to': self.account.address,
+                        'value': 0,
+                        'nonce': nonce,
+                        'gas': 21000,
+                        'gasPrice': self.web3.eth.gas_price * 2,  # Double the current gas price
+                        'chainId': self.chain_id  # Add chain ID for EIP-155
+                    }
+
+                    signed_tx = self.web3.eth.account.sign_transaction(replacement_tx, self.account.key)
+                    try:
+                        self.web3.eth.send_raw_transaction(signed_tx.rawTransaction)
+                        vana.logging.info(f"Sent replacement transaction for nonce {nonce}")
+                    except Exception as e:
+                        vana.logging.warning(f"Failed to replace transaction with nonce {nonce}: {str(e)}")
+
+                # Wait for transactions to be processed
+                time.sleep(30)
```

Comment:
Improvement for the future:

This looks strange to me, we don't know how long to wait (it could be 5 txns, or 500). If the goal is to wait for all the pending transactions to complete, can we just poll on the status of the last transaction (highest nonce), or a max_timeout, whichever comes first?

Reply:
Great point about polling the last transaction status! I'll create a separate PR for this improvement.

I want to test this change carefully, but currently our test environment has no stuck transactions to verify against. Will work on setting up a reliable test scenario that can consistently reproduce the pending transaction buildup.